### ISSUE TITLE
train_and_evaluate does not preserve the Dataset iterator state across train/eval

### ISSUE THREAD

**Issue by superbobry** [tensorflow/tensorflow#19062]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: N/A
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.8.0-0-g93bc2e2072 1.8.0
- **Python version**: 3.6
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: see below

### Describe the problem

When the input function is based on the one-shot dataset iterator, the training phase always starts from the beginning of the iterator. That is, the iterator state gets reset in between the train/eval phases. Therefore, if the dataset is big enough, then the training would only see a subset of the data, which can be processed in `eval_spec.throttle_secs`. 

I think the issue is caused by the fact that the graph is persisted before transitioning to the next phase, and restored upon reentering training. However, I find the behaviour a bit counterintuitive, so if it is not a bug, it should be mentioned in the `train_and_evaluate` docs.

### Source code / logs

Here is a small example demonstrating the issue:

```python
import tensorflow as tf


def input_fn(data):
    dataset = tf.data.Dataset.from_tensor_slices(data)
    dataset = dataset.batch(batch_size=1)
    x = dataset.make_one_shot_iterator().get_next()
    return {"x": tf.Print(x, [x])}, x


if __name__ == "__main__":
    model = tf.estimator.LinearRegressor(feature_columns=[
        tf.feature_column.numeric_column("x")
    ])

    train_spec = tf.estimator.TrainSpec(
        input_fn=lambda: input_fn(list(range(2**20))),
        max_steps=2)
    eval_spec = tf.estimator.EvalSpec(
        input_fn=lambda: input_fn([42]),
        steps=1,
        start_delay_secs=1,
        throttle_secs=1)

    tf.logging.set_verbosity("INFO")
    tf.train.create_global_step()
    tf.estimator.train_and_evaluate(model, train_spec, eval_spec)
```

The code produces the following log output (I've omitted irrelevant lines):

```
INFO:tensorflow:Start train and evaluate loop. The evaluate will happen after 1 secs (eval_spec.throttle_secs) or training is finished.
INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Create CheckpointSaverHook.
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
[0]
INFO:tensorflow:Saving checkpoints for 1 into /var/folders/wr/s7brqkzj74v4pmdwkwn19321l34xg2/T/tmpgtq1ap9_/model.ckpt.
INFO:tensorflow:loss = 0.0, step = 1
INFO:tensorflow:Loss for final step: 0.0.
...
INFO:tensorflow:Starting evaluation at 2018-05-03-16:21:51
...
INFO:tensorflow:Finished evaluation at 2018-05-03-16:21:51
...
INFO:tensorflow:Restoring parameters from /var/folders/wr/s7brqkzj74v4pmdwkwn19321l34xg2/T/tmpgtq1ap9_/model.ckpt-1
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
[0]
INFO:tensorflow:Saving checkpoints for 2 into /var/folders/wr/s7brqkzj74v4pmdwkwn19321l34xg2/T/tmpgtq1ap9_/model.ckpt.
INFO:tensorflow:loss = 0.0, step = 2
INFO:tensorflow:Loss for final step: 0.0.
...
```

**Comment by JuanjoAlegria** [comment]
Just today I came across this problem, and I can't think of a workaround. Also, I think this is also related to #13895.

**Comment by superbobry** [comment]
@JuanjoAlegria the workaround is to run the TF cluster on a single machine with just two tasks: chief doing the training, and evaluator doing solely the evaluation. Each task should be started in a separate process, e.g. via the [`ProcessPoolExecutor`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor).

**Comment by JuanjoAlegria** [comment]
Is there any news about this issue? I agree with @superbobry that, if this is not a bug, at least is an issue that should be highlighted in the docs.

Thanks!

**Comment by bhack** [comment]
This caused also problems to the cache. Check https://github.com/tensorflow/tensorflow/issues/18266

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ispirmustafa** [comment]
Hi @superbobry 
We have updated the behavior of train_and_evaluate. Now train is called only once. So you should not have cache problem in train input-fn.
Having said that eval input-fn will be called for every evaluate.

**Comment by bhack** [comment]
@ispirmustafa So also  https://github.com/tensorflow/tensorflow/issues/18266 is it solved?

**Comment by bhack** [comment]
@ipirmustafa is there a plan also for evaluate?

**Comment by guillaumekln** [comment]
Thanks @ispirmustafa, that is a welcome change. Looks like this documentation should be updated then:

https://github.com/tensorflow/tensorflow/blob/3edb609926f2521c726737fc1efeae1572dc6581/tensorflow/python/estimator/training.py#L280-L284

**Comment by superbobry** [comment]
Thanks for looking into this, @ispirmustafa! Am I right that the training graph is no longer persisted in local mode? If not, how did you solve iterator persistence issue? 

Also, for future reference, here's the commit fixing the train part: 3edb609926f2521c726737fc1efeae1572dc6581.

**Comment by ispirmustafa** [comment]
Hi @guillaumekln , Yes that should be updated. we'll do that

**Comment by ispirmustafa** [comment]
Hi @bhack ,
For evaluate we don't have any plan to change graph re-creation.

**Comment by ispirmustafa** [comment]
Hi @superbobry, 
I guess you mean whether we keep same graph with multiple train graph by 'persisting'.
No we're not doing that. We call train only once. We do evaluation via a hook (listener)

**Comment by superbobry** [comment]
@ispirmustafa yes, thanks for clarifying this. I think we can close the issue once the documentation clearly reflects the evaluation graph re-creation.

One final question: is there a reason for not caching the evaluation graph in the hook? i.e. do we really need to recreate it?

**Comment by ispirmustafa** [comment]
Keeping evaluation graph needs to hijack existing estimator API.

**Comment by superbobry** [comment]
I'm sorry, I think I miss context to understand what needs to be hijacked.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ispirmustafa: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by formigone** [comment]
Am I  understanding this right? Suppose my input_fn returns a `tf.data.TextLineDataset(filename).map (lamda line: tf.decode_csv(line))` that references a csv with, say, 1,000,000 lines, and it shuffles with `dataset.shuffle(buffer_size=256)`. If the evaluation runs after a few hundred steps of training, when training resumes, will the input_fn used in training start reading thr cvs from where it left off before doing the evaluation, or will it start over from the first line?

**Comment by ispirmustafa** [comment]
@formigone your statement is correct. train input_fn will not be impacted by eval. this is the case for tf.version>=1.9
